pypose/module/cameras.py

```diff
@@ -0,0 +1,149 @@
+        """
+        raise NotImplementedError
+
+    def reprojection_error(self, pts_w, img_pts):
+        """
+        Calculate the reprojection error.
+        Args:
+            pts_w: The object points in world coordinate. The shape is (..., N, 3).
+            img_pts: The image points. The shape is (..., N, 2).
+        Returns:
+            error: The reprojection error. The shape is (..., ).
+        """
+        # Calculate the image points
+        img_repj = self.full_projection_transform(pts_w)
+
+        error = torch.linalg.norm(img_repj - img_pts, dim=-1)
+        error = torch.mean(error, dim=-1)
+
+        return error
+
+
+class PerspectiveCameras(CamerasBase):
+    """
+    A class which stores a batch of parameters to generate a batch of
+    transformation matrices using the multi-view geometry convention for
+    perspective camera.
+
+    Args:
+        pose (LieTensor): A tensor of shape (B, *) where B is the batch size.
+        intrinsics (Optional): A projection matrix of shape (N, 3, 3)
+
+    Examples:
+        >>> import torch
+        >>> import pypose
+        >>> # create some random data
+        >>> R = torch.tensor([[ 0.70710678,  0., -0.70710678], [ 0. , 1., 0. ],[ 0.70710678, 0.,  0.70710678]])
+        >>> t = torch.tensor([0., -8., 0.])
+        >>> f = 2
+        >>> img_size = (7, 7)
+        >>> projection_matrix = torch.tensor([[f, 0, img_size[0] / 2,], [0, f, img_size[1] / 2,], [0, 0, 1, ]])
+        >>> pts_w = torch.tensor([[ 2.8284,  8.0000,  0.0000], [ 2.1213,  8.0000,  0.7071], [ 0.7071,  9.0000,  0.7071], [ 0.7071,  8.0000,  0.7071], [ 5.6569, 13.0000, -1.4142]])
+        >>> # instantiate the camera
+        >>> Rt = torch.cat((R, t.unsqueeze(-1)), dim=-1)
+        >>> camera = pypose.module.PerspectiveCameras(pose=pypose.mat2SE3(Rt), intrinsics=projection_matrix)
+        >>> # transform the points to image coordinates
+        >>> img_pts = camera.full_projection_transform(pts_w)
+        >>> img_pts
+        tensor([[5.5000, 3.5000],
+                [4.5000, 3.5000],
+                [3.5000, 5.5000],
+                [3.5000, 3.5000],
+                [6.8333, 6.8333]])
+    """
+
+    def __init__(
+            self,
+            pose: Optional[pypose.LieTensor] = None,
+            intrinsics: Optional[torch.Tensor] = None,
+    ):
+        super().__init__(pose=pose)
+        if intrinsics is not None:
+            self.intrinsics = intrinsics
+        elif self.intrinsics is None:
+            self.intrinsics = torch.eye(3)
+
+    def projection_transform(self, points):
+        """
+        Transform camera view coordinate points to pixel locations.
+        Args:
+            points (torch.Tensor): A tensor of shape (B, N, 3) where B is the batch size.
+        Returns:
+            points (torch.Tensor): A tensor of shape (B, N, 2) where B is the batch size.
+        """
+        img_repj = torch.matmul(points, self.intrinsics.transpose(-2, -1))
```

Comment:
do you want
```python
img_repj = self.intrinsics @ points
```
or
```python
img_repj = self.intrinsics.mT @ points
```

Reply:
I wrote `points.matmul(self.intrinsics.transpose(-2, -1))` because it would be the cleanest way of writing it. Left matmul of intrinsics has some problem with broadcasting

```
self.intrinsics.shape
PyDev console: starting.
torch.Size([1, 3, 3])
points.shape
torch.Size([1, 5, 3])
pypose.bmv(self.intrinsics.unsqueeze(-2), points)
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
  File "/Users/zitongzhan/Documents/pypose/pypose/basics/ops.py", line 34, in bmv
    return torch.matmul(mat, vec.unsqueeze(-1), out=out).squeeze_(-1)
RuntimeError: The size of tensor a (3) must match the size of tensor b (5) at non-singleton dimension 1
```

Reply:
changing it to `img_repj = points.matmul(self.intrinsics.mT)`